### PR TITLE
[FEATURE]Check date type to avoid scan all partitions

### PR DESCRIPTION
**Pull request by xinghuayu007**
## Proposed changes

select day from test where day='2020-10-32', table 'test' is parititioned by day. In this case, '2020-10-32' will be taken as CastExpr not LiteralExpr, and condition "day='2020-10-32'" will not be recognized as partitionfilter. This case will scan all partitions. To avoid scall all partitions, it is better to filter invalid date value.

issue: #4755
## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [] Bugfix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] Documentation Update (if none of the other choices apply)
- [] Code refactor (Modify the code structure, format the code, etc...)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have create an issue on (Fix #ISSUE), and have described the bug/feature there in detail
- [x] Compiling and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] If this change need a document change, I have updated the document
- [x] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/ExprRewriter.java [hunk 1]
```diff
@@ -55,6 +58,12 @@ public Expr rewrite(Expr expr, Analyzer analyzer) throws AnalysisException {
                 rewrittenExpr = applyRuleRepeatedly(rewrittenExpr, rule, analyzer);
             }
         } while (oldNumChanges != numChanges_);
+        if (expr instanceof BinaryPredicate) {
```
Comment (xy720) [inline fe/fe-core/src/main/java/org/apache/doris/rewrite/ExprRewriter.java:61]: Why not check the type in analysis state?

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/ExprRewriter.java [hunk 2]
```diff
@@ -55,6 +58,12 @@ public Expr rewrite(Expr expr, Analyzer analyzer) throws AnalysisException {
...
+            Expr valueExpr = ((BinaryPredicate) expr).getBinding();
+            if(valueExpr != null && valueExpr.getType() == Type.DATETIME && valueExpr instanceof CastExpr) {
+                throw new AnalysisException("invalid date type :" + valueExpr.toSql());
```
Comment (xy720) [inline fe/fe-core/src/main/java/org/apache/doris/rewrite/ExprRewriter.java:64]: I think throwing a exception is not the best solution of this problem, may be rewrite the expr `day = '2020-10-32'` to something like `1 = 0` in where statement is better. We should discuss it first.

Comment (xinghuayu007) [reply fe/fe-core/src/main/java/org/apache/doris/rewrite/ExprRewriter.java:64]: > I think throwing a exception is not the best solution of this problem, may be rewrite the binary expr `day = '2020-10-32'` to something like contant expr BoolLiteral[false] in where statement is better. We should discuss it first.

ok，I will modify it;

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/BinaryPredicatesDateRule.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ * date = "2020-10-32" => FALSE
+ */
+public class BinaryPredicatesDateRule implements ExprRewriteRule {
+    public static ExprRewriteRule INSTANCE = new BinaryPredicatesDateRule();
+
+    @Override
+    public Expr apply(Expr expr, Analyzer analyzer) throws AnalysisException {
+        if (!(expr instanceof BinaryPredicate)) return expr;
+        Expr valueExpr = expr.getChild(1);
+        if(valueExpr.getType() == Type.DATETIME && valueExpr instanceof CastExpr) {
```
Comment (xy720) [inline fe/fe-core/src/main/java/org/apache/doris/rewrite/BinaryPredicatesDateRule.java:41]:
```suggestion
        if(valueExpr.getType() == Type.DATETIME && valueExpr instanceof CastExpr && valueExpr.getChild(0).getType() == Type.VARCHAR) {
```
There is something wrong with this If statement.
For example, `where day = 20201030` will rewrite to `where false`, which is not right.
You should only support covert StringLiteral now:
`where day = '20201032'` -> `where false`

**Review by xy720 — COMMENTED** [review]
LGTM

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/BinaryPredicatesDateRule.java (hunk 3 above)
Comment (xinghuayu007) [reply fe/fe-core/src/main/java/org/apache/doris/rewrite/BinaryPredicatesDateRule.java:41]: The rule of FoldConstantsRule will convert `where day = 20201030` into `where day='2020-10-30 00:00:00'`, whic is DateLiteral type.

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**